File: Data Structures & Algorithms/max-water-container/submission-1.py
class Solution:
    def maxArea(self, heights: List[int]) -> int:
        l, r = 0, len(heights) - 1
        res = 0
        while l < r:
            minHeight = min(heights[l], heights[r])
            water = (r-l) * minHeight
            print(water)
            res = max(res, water)
            if heights[l] == minHeight:
                l += 1
            else:
                r -= 1
        return res

File: Data Structures & Algorithms/trapping-rain-water/submission-1.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int length = height.size();
        vector <int> max_r(length+2, 0);
        vector <int> max_l(length+2, 0);
        int res = 0;

        for (int i = 0; i < length; i++) max_l[i+1] = max(max_l[i], height[i]);
        for (int i = length; i > 0; i--) max_r[i] = max(max_r[i+1], height[i-1]);

        for (int i = 0; i < length; i++) res += min(max_l[i+1], max_r[i+1]) - height[i];
        return res;
    }
};
